# BS_bias

Source code for the manuscript "Comparison of whole-genome bisulfite sequencing library preparation strategies identifies sources of biases affecting DNA methylation data". 

Abstract

Background: Whole-genome bisulfite sequencing (WGBS) is becoming an increasingly accessible technique, used
widely for both fundamental and disease-oriented research. Library preparation methods benefit from a variety of
available kits, polymerases and bisulfite conversion protocols. Although some steps in the procedure, such as PCR
amplification, are known to introduce biases, a systematic evaluation of biases in WGBS strategies is missing.

Results: We perform a comparative analysis of several commonly used pre- and post-bisulfite WGBS library
preparation protocols for their performance and quality of sequencing outputs. Our results show that
bisulfite conversion per se is the main trigger of pronounced sequencing biases, and PCR amplification
builds on these underlying artefacts. The majority of standard library preparation methods yield a significantly biased
sequence output and overestimate global methylation. Importantly, both absolute and relative methylation levels at
specific genomic regions vary substantially between methods, with clear implications for DNA methylation studies.

Conclusions: We show that amplification-free library preparation is the least biased approach for WGBS. In protocols
with amplification, the choice of bisulfite conversion protocol or polymerase can significantly minimize artefacts. To aid
with the quality assessment of existing WGBS datasets, we have integrated a bias diagnostic tool in the
Bismark package and offer several approaches for consideration during the preparation and analysis of WGBS datasets.

Link to manuscript: http://rdcu.be/JaqP
